core/trino-main/src/main/java/io/trino/cost/CachingTableStatsProvider.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.cost;
+
+import io.trino.Session;
+import io.trino.metadata.Metadata;
+import io.trino.metadata.TableHandle;
+import io.trino.spi.statistics.TableStatistics;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class CachingTableStatsProvider
+        implements TableStatsProvider
+{
+    private final Metadata metadata;
+    private final Map<TableHandle, TableStatistics> cache = new HashMap<>();
```

Comment:
A connector creates new TableHandles during various `ConnectorMetadata.apply*` calls.
A table handle may become "seen" in the plan and then discarded, make obsolete.

I think we should use weak keys here. Otherwise we need to size this as a regular cache.
(`WeakHashMap` provides weak keys with equality-based lookup, so i'd recommend that)

Reply:
I don't fully understand this comment.. My understanding is that `CachingTableStatsProvider` is created per query and its lifecycle is within query planning. So the cache here can be GCed after the query planning. What issue did you see with it?